File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/event/registration/DefaultEndpointRegistrationManager.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.event.registration;

import org.kaaproject.kaa.client.channel.ProfileTransport;
import org.kaaproject.kaa.client.channel.UserTransport;
import org.kaaproject.kaa.client.context.ExecutorContext;
import org.kaaproject.kaa.client.event.EndpointAccessToken;
import org.kaaproject.kaa.client.event.EndpointKeyHash;
import org.kaaproject.kaa.client.persistence.KaaClientState;
import org.kaaproject.kaa.common.endpoint.gen.EndpointAttachResponse;
import org.kaaproject.kaa.common.endpoint.gen.EndpointDetachResponse;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponseResultType;
import org.kaaproject.kaa.common.endpoint.gen.UserAttachNotification;
import org.kaaproject.kaa.common.endpoint.gen.UserAttachRequest;
import org.kaaproject.kaa.common.endpoint.gen.UserAttachResponse;
import org.kaaproject.kaa.common.endpoint.gen.UserDetachNotification;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Default {@link EndpointRegistrationManager} implementation.
 *
 * @author Taras Lemkin
 */
public class DefaultEndpointRegistrationManager implements EndpointRegistrationManager,
        EndpointRegistrationProcessor {

  private static final Logger LOG = LoggerFactory.getLogger(
          DefaultEndpointRegistrationManager.class);
  private static final Random RANDOM = new Random();

  private final KaaClientState state;
  private final Map<Integer, OnAttachEndpointOperationCallback> endpointAttachListeners =
          new ConcurrentHashMap<>();
  private final Map<Integer, OnDetachEndpointOperationCallback> endpointDetachListeners =
          new ConcurrentHashMap<>();

  private final Map<Integer, EndpointAccessToken> attachEndpointRequests =
          new ConcurrentHashMap<>();
  private final Map<Integer, EndpointKeyHash> detachEndpointRequests = new ConcurrentHashMap<>();
  private final ExecutorContext executorContext;
  private UserAttachRequest userAttachRequest;
  private UserAttachCallback userAttachCallback;
  private AttachEndpointToUserCallback attachEndpointToUserCallback;
  private DetachEndpointFromUserCallback detachEndpointFromUserCallback;
  private volatile UserTransport userTransport;
  private volatile ProfileTransport profileTransport;

  /**
   * All-args constructor.
   */
  public DefaultEndpointRegistrationManager(KaaClientState state, ExecutorContext executorContext,
                                            UserTransport userTransport,
                                            ProfileTransport profileTransport) {
    this.userTransport = userTransport;
    this.profileTransport = profileTransport;
    this.state = state;
    this.executorContext = executorContext;
    String endpointAccessToken = state.getEndpointAccessToken();
    if (endpointAccessToken == null || endpointAccessToken.length() == 0) {
      state.refreshEndpointAccessToken();
    }
  }

  private String regenerateEndpointAccessToken() {
    String newEndpointAccessToken = state.refreshEndpointAccessToken();
    LOG.info("New endpoint access token is generated: '{}'", newEndpointAccessToken);
    onEndpointAccessTokenChanged();
    return newEndpointAccessToken;
  }


  /**
   * Updates an endpoint access token.
   *
   * @param token new endpoint access token
   */
  public void updateEndpointAccessToken(String token) {
    state.setEndpointAccessToken(token);
    state.setIfNeedProfileResync(true);
    onEndpointAccessTokenChanged();
  }

  public String refreshEndpointAccessToken() {
    return regenerateEndpointAccessToken();
  }

  @Override
  public void attachEndpoint(EndpointAccessToken endpointAccessToken,
                             OnAttachEndpointOperationCallback resultListener) {
    int requestId = getRandomInt();
    LOG.info("Going to attach Endpoint by access token: {}", endpointAccessToken);
    attachEndpointRequests.put(requestId, endpointAccessToken);
    if (resultListener != null) {
      endpointAttachListeners.put(requestId, resultListener);
    }
    if (userTransport != null) {
      userTransport.sync();
    }
  }

  @Override
  public void detachEndpoint(
          EndpointKeyHash endpointKeyHash, OnDetachEndpointOperationCallback resultListener) {
    int requestId = getRandomInt();
    LOG.info("Going to detach Endpoint by endpoint key hash: {}", endpointKeyHash);
    detachEndpointRequests.put(requestId, endpointKeyHash);
    if (resultListener != null) {
      endpointDetachListeners.put(requestId, resultListener);
    }
    if (userTransport != null) {
      userTransport.sync();
    }
  }

  @Override
  public void attachUser(
          String userExternalId, String userAccessToken, UserAttachCallback callback) {
    if (UserVerifierConstants.DEFAULT_USER_VERIFIER_TOKEN != null) {
      attachUser(UserVerifierConstants.DEFAULT_USER_VERIFIER_TOKEN, userExternalId,
              userAccessToken, callback);
    } else {
      throw new IllegalStateException(
              "Default user verifier was not defined during SDK generation process!");
    }
  }

  @Override
  public void attachUser(String userVerifierToken, String userExternalId, String userAccessToken,
                         UserAttachCallback callback) {
    userAttachRequest = new UserAttachRequest(userVerifierToken, userExternalId, userAccessToken);
    userAttachCallback = callback;
    if (userTransport != null) {
      userTransport.sync();
    }
  }

  public Map<EndpointAccessToken, EndpointKeyHash> getAttachedEndpointList() {
    return state.getAttachedEndpointsList();
  }

  @Override
  public void onUpdate(List<EndpointAttachResponse> attachResponses,
                       List<EndpointDetachResponse> detachResponses,
                       final UserAttachResponse userResponse,
                       final UserAttachNotification userAttachNotification,
                       final UserDetachNotification userDetachNotification) throws IOException {
    if (userResponse != null) {
      if (userAttachCallback != null) {
        final UserAttachCallback callback = userAttachCallback;
        executorContext.getCallbackExecutor().submit(new Runnable() {
          @Override
          public void run() {
            callback.onAttachResult(userResponse);

          }
        });
        userAttachCallback = null;
      }
      if (userResponse.getResult() == SyncResponseResultType.SUCCESS) {
        state.setAttachedToUser(true);
        if (attachEndpointToUserCallback != null) {
          final AttachEndpointToUserCallback callback = attachEndpointToUserCallback;
          executorContext.getCallbackExecutor().submit(new Runnable() {
            @Override
            public void run() {
              callback.onAttachedToUser(userAttachRequest.getUserExternalId(),
                      state.getEndpointAccessToken());
            }
          });
        }
      }
      userAttachRequest = null;
    }

    if (attachResponses != null && !attachResponses.isEmpty()) {
      for (EndpointAttachResponse attached : attachResponses) {
        notifyAttachedListener(attached.getResult(),
                endpointAttachListeners.remove(attached.getRequestId()), new EndpointKeyHash(
            attached.getEndpointKeyHash()));
        attachEndpointRequests.remove(attached.getRequestId());
      }
    }
    if (detachResponses != null && !detachResponses.isEmpty()) {
      for (EndpointDetachResponse detached : detachResponses) {
        notifyDetachedListener(detached.getResult(),
                endpointDetachListeners.remove(detached.getRequestId()));
        EndpointKeyHash endpointKeyHash = detachEndpointRequests.remove(detached.getRequestId());
        if (endpointKeyHash != null && detached.getResult() == SyncResponseResultType.SUCCESS) {
          if (endpointKeyHash.equals(state.getEndpointKeyHash())) {
            state.setAttachedToUser(false);
          }
        }
      }
    }

    if (userAttachNotification != null) {
      state.setAttachedToUser(true);
      if (attachEndpointToUserCallback != null) {
        final AttachEndpointToUserCallback callback = attachEndpointToUserCallback;
        executorContext.getCallbackExecutor().submit(new Runnable() {
          @Override
          public void run() {
            callback.onAttachedToUser(userAttachNotification.getUserExternalId(),
                userAttachNotification.getEndpointAccessToken());
          }
        });
      }
    }

    if (userDetachNotification != null) {
      state.setAttachedToUser(false);
      if (detachEndpointFromUserCallback != null) {
        final DetachEndpointFromUserCallback callback = detachEndpointFromUserCallback;
        executorContext.getCallbackExecutor().submit(new Runnable() {
          @Override
          public void run() {
            callback.onDetachedFromUser(userDetachNotification.getEndpointAccessToken());
          }
        });
      }
    }
  }

  private void notifyAttachedListener(final SyncResponseResultType result,
                                      final OnAttachEndpointOperationCallback operationCallback,
                                      final EndpointKeyHash keyHash) {
    if (operationCallback != null) {
      executorContext.getCallbackExecutor().submit(new Runnable() {
        @Override
        public void run() {
          operationCallback.onAttach(result, keyHash);
        }
      });
    }
  }

  private void notifyDetachedListener(final SyncResponseResultType result,
                                      final OnDetachEndpointOperationCallback operationCallback) {
    if (operationCallback != null) {
      executorContext.getCallbackExecutor().submit(new Runnable() {
        @Override
        public void run() {
          operationCallback.onDetach(result);
        }
      });
    }
  }

  private void onEndpointAccessTokenChanged() {
    if (profileTransport != null) {
      profileTransport.sync();
    }
  }

  @Override
  public Map<Integer, EndpointAccessToken> getAttachEndpointRequests() {
    Map<Integer, EndpointAccessToken> result = new HashMap<>();
    synchronized (attachEndpointRequests) {
      result.putAll(attachEndpointRequests);
    }
    return result;
  }

  @Override
  public Map<Integer, EndpointKeyHash> getDetachEndpointRequests() {
    Map<Integer, EndpointKeyHash> result = new HashMap<>();
    synchronized (detachEndpointRequests) {
      result.putAll(detachEndpointRequests);
    }
    return result;
  }

  @Override
  public UserAttachRequest getUserAttachRequest() {
    return userAttachRequest;
  }

  @Override
  public boolean isAttachedToUser() {
    return state.isAttachedToUser();
  }

  @Override
  public void setAttachedCallback(AttachEndpointToUserCallback listener) {
    this.attachEndpointToUserCallback = listener;
  }

  @Override
  public void setDetachedCallback(DetachEndpointFromUserCallback listener) {
    this.detachEndpointFromUserCallback = listener;
  }

  private synchronized int getRandomInt() {
    return RANDOM.nextInt();
  }
}

File: client/client-multi/client-objective-c/Kaa/DefaultEndpointRegistrationManager.m
/*
 * Copyright 2014-2015 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "DefaultEndpointRegistrationManager.h"
#import "EventDelegates.h"
#import "EndpointGen.h"
#import "UserVerifierConstants.h"
#include <stdlib.h>
#import "KaaLogging.h"
#import "KaaExceptions.h"

#define TAG @"DefaultEndpointRegistrationManager >>>"

@interface DefaultEndpointRegistrationManager ()

@property (nonatomic,strong) id<KaaClientState> state;
@property (nonatomic,strong) id<ExecutorContext> context;
@property (nonatomic,strong) volatile id<UserTransport> userTransport;
@property (nonatomic,strong) volatile id<ProfileTransport> profileTransport;

@property (nonatomic,weak) id<AttachEndpointToUserDelegate> attachEndpointToUserDelegate;
@property (nonatomic,weak) id<DetachEndpointFromUserDelegate> detachEndpointFromUserDelegate;

@property (nonatomic,strong) UserAttachRequest *userAttachRequest;
@property (nonatomic,weak) id<UserAttachDelegate> userAttachDelegate;

@property (nonatomic,strong) NSMutableDictionary *endpointAttachDelegates;
@property (nonatomic,strong) NSMutableDictionary *endpointDetachDelegates;

@property (nonatomic,strong) NSMutableDictionary *attachEndpointRequests;
@property (nonatomic,strong) NSMutableDictionary *detachEndpointRequests;

- (void)onEndpointAccessTokenChanged;
- (void)notifyAttachedDelegate:(SyncResponseResultType)result
                      delegate:(id<OnAttachEndpointOperationDelegate>)delegate
                       keyHash:(EndpointKeyHash *)keyHash;
- (void)notifyDetachedDelegate:(SyncResponseResultType)result
                      delegate:(id<OnDetachEndpointOperationDelegate>)delegate;
- (int)getRandomInteger;

@end

@implementation DefaultEndpointRegistrationManager

- (instancetype)initWith:(id<KaaClientState>)state
         executorContext:(id<ExecutorContext>)context
           userTransport:(id<UserTransport>)userTransport
        profileTransport:(id<ProfileTransport>)profileTransport {
    self = [super init];
    if (self) {
        self.state = state;
        self.context = context;
        self.userTransport = userTransport;
        self.profileTransport = profileTransport;
        
        self.endpointAttachDelegates = [NSMutableDictionary dictionary];
        self.endpointDetachDelegates = [NSMutableDictionary dictionary];
        self.attachEndpointRequests = [NSMutableDictionary dictionary];
        self.detachEndpointRequests = [NSMutableDictionary dictionary];
        
        NSString *endpointAccessToken = [self.state endpointAccessToken];
        if (!endpointAccessToken || endpointAccessToken.length == 0) {
            [self.state refreshEndpointAccessToken];
        }
    }
    return self;
}

- (void)updateEndpointAccessToken:(NSString *)token {
    [self.state setEndpointAccessToken:token];
}

- (NSString *)refreshEndpointAccessToken {
    NSString *newEndpointAccessToken = [self.state refreshEndpointAccessToken];
    DDLogInfo(@"%@ New endpoint access token is generated: %@", TAG, newEndpointAccessToken);
    [self onEndpointAccessTokenChanged];
    return newEndpointAccessToken;
}

- (NSDictionary *)getAttachedEndpointList {
    return [self.state attachedEndpoints];
}

- (void)attachEndpoint:(EndpointAccessToken *)accessToken delegate:(id<OnAttachEndpointOperationDelegate>)delegate {
    NSNumber *requestId = [NSNumber numberWithInt:[self getRandomInteger]];
    DDLogInfo(@"%@ Going to attach Endpoint by access token: %@", TAG, accessToken);
    @synchronized (self.attachEndpointRequests) {
        [self.attachEndpointRequests setObject:accessToken forKey:requestId];
    }
    if (delegate) {
        @synchronized (self.endpointAttachDelegates) {
            [self.endpointAttachDelegates setObject:delegate forKey:requestId];
        }
    }
    if (self.userTransport) {
        [self.userTransport sync];
    }
}

- (void)detachEndpoint:(EndpointKeyHash *)keyHash delegate:(id<OnDetachEndpointOperationDelegate>)delegate {
    NSNumber *requestId = [NSNumber numberWithInt:[self getRandomInteger]];
    DDLogInfo(@"%@ Going to detach Endpoint by endpoint key hash: %@", TAG, keyHash);
    @synchronized (self.detachEndpointRequests) {
        [self.detachEndpointRequests setObject:keyHash forKey:requestId];
    }
    if (delegate) {
        @synchronized (self.endpointDetachDelegates) {
            [self.endpointDetachDelegates setObject:delegate forKey:requestId];
        }
    }
    if (self.userTransport) {
        [self.userTransport sync];
    }
}

- (void)attachUser:(NSString *)userExternalId userAccessToken:(NSString *)token delegate:(id<UserAttachDelegate>)delegate {
    if (DEFAULT_USER_VERIFIER_TOKEN != nil) {
        [self attachUser:DEFAULT_USER_VERIFIER_TOKEN userExternalId:userExternalId userAccessToken:token delegate:delegate];
    } else {
        [NSException raise:KaaIllegalStateException format:@"Default user verifier was not defined during SDK generation process!"];
    }
}

- (void)attachUser:(NSString *)userVerifierToken
    userExternalId:(NSString *)externalId
   userAccessToken:(NSString *)token
          delegate:(id<UserAttachDelegate>)delegate {
    self.userAttachRequest = [[UserAttachRequest alloc] init];
    self.userAttachRequest.userVerifierId = userVerifierToken;
    self.userAttachRequest.userExternalId = externalId;
    self.userAttachRequest.userAccessToken = token;
    
    self.userAttachDelegate = delegate;
    if (self.userTransport) {
        [self.userTransport sync];
    }
}

- (void)onUpdate:(NSArray *)attachResponses detachResponses:(NSArray *)detachResponses
    userResponse:(UserAttachResponse *)userResponse
userAttachNotification:(UserAttachNotification *)attachNotification
userDetachNotification:(UserDetachNotification *)detachNotification {
    
    if (userResponse) {
        if (self.userAttachDelegate) {
            __block id<UserAttachDelegate> delegate = self.userAttachDelegate;
            __weak typeof(self)weakSelf = self;
            [[self.context getCallbackExecutor] addOperationWithBlock:^{
                [delegate onAttachResult:userResponse];
            }];
            weakSelf.userAttachDelegate = nil;
        }
        if (userResponse.result == SYNC_RESPONSE_RESULT_TYPE_SUCCESS) {
            [self.state setIsAttachedToUser:YES];
            if (self.attachEndpointToUserDelegate && self.userAttachRequest) {
                __block UserAttachRequest *request = self.userAttachRequest;
                __weak typeof(self)weakSelf = self;
                [[self.context getCallbackExecutor] addOperationWithBlock:^{
                    [weakSelf.attachEndpointToUserDelegate onAttachedToUser:request.userExternalId token:[weakSelf.state endpointAccessToken]];
                }];
            }
        }
        self.userAttachRequest = nil;
    }
    
    if (attachResponses && [attachResponses count] > 0) {
        for (EndpointAttachResponse *attached in attachResponses) {
            @synchronized (self.endpointAttachDelegates) {
                id<OnAttachEndpointOperationDelegate> delegate = [self.endpointAttachDelegates objectForKey: [NSNumber numberWithInt:attached.requestId]];
                [self.endpointAttachDelegates removeObjectForKey:[NSNumber numberWithInt:attached.requestId]];
                EndpointKeyHash *keyHash = [[EndpointKeyHash alloc] initWithKeyHash:attached.endpointKeyHash.data];
                [self notifyAttachedDelegate:attached.result delegate:delegate keyHash:keyHash];
            }
            @synchronized (self.attachEndpointRequests) {
                [self.attachEndpointRequests removeObjectForKey:[NSNumber numberWithInt:attached.requestId]];
            }
        }
    }
    
    if (detachResponses && [detachResponses count] > 0) {
        for (EndpointDetachResponse *detached in detachResponses) {
            @synchronized (self.endpointDetachDelegates) {
                NSNumber *requestId = [NSNumber numberWithInt:detached.requestId];
                id<OnDetachEndpointOperationDelegate> delegate = [self.endpointDetachDelegates objectForKey: requestId];
                [self.endpointDetachDelegates removeObjectForKey:[NSNumber numberWithInt:detached.requestId]];
                [self notifyDetachedDelegate:detached.result delegate:delegate];
            }
            EndpointKeyHash *keyHash = nil;
            @synchronized (self.detachEndpointRequests) {
                keyHash = [self.detachEndpointRequests objectForKey:[NSNumber numberWithInt:detached.requestId]];
                [self.detachEndpointRequests removeObjectForKey:[NSNumber numberWithInt:detached.requestId]];
            }
            if (keyHash && detached.result == SYNC_RESPONSE_RESULT_TYPE_SUCCESS) {
                if ([keyHash isEqual:[self.state endpointKeyHash]]) {
                    [self.state setIsAttachedToUser:NO];
                }
            }
        }
    }
    if (attachNotification) {
        [self.state setIsAttachedToUser:YES];
        if (self.attachEndpointToUserDelegate) {
            __weak typeof(self)weakSelf = self;
            [[self.context getCallbackExecutor] addOperationWithBlock:^{
                [weakSelf.attachEndpointToUserDelegate onAttachedToUser:attachNotification.userExternalId
                                                                  token:attachNotification.endpointAccessToken];
            }];
        }
    }
    if (detachNotification) {
        [self.state setIsAttachedToUser:NO];
        if (self.detachEndpointFromUserDelegate) {
            __weak typeof(self)weakSelf = self;
            [[self.context getCallbackExecutor] addOperationWithBlock:^{
                [weakSelf.detachEndpointFromUserDelegate onDetachedFromUser:detachNotification.endpointAccessToken];
            }];
        }
    }
}

- (NSDictionary *)getAttachEndpointRequests {
    @synchronized (self.attachEndpointRequests) {
        return [NSMutableDictionary dictionaryWithDictionary:self.attachEndpointRequests];
    }
}

- (NSDictionary *)getDetachEndpointRequests {
    @synchronized (self.detachEndpointRequests) {
        return [NSMutableDictionary dictionaryWithDictionary:self.detachEndpointRequests];
    }
}

- (UserAttachRequest *)getUserAttachRequest {
    return _userAttachRequest;
}

- (BOOL)isAttachedToUser {
    return [self.state isAttachedToUser];
}

- (void)setAttachedDelegate:(id<AttachEndpointToUserDelegate>)delegate {
    self.attachEndpointToUserDelegate = delegate;
}

- (void)setDetachedDelegate:(id<DetachEndpointFromUserDelegate>)delegate {
    self.detachEndpointFromUserDelegate = delegate;
}

- (void)notifyAttachedDelegate:(SyncResponseResultType)result
                      delegate:(id<OnAttachEndpointOperationDelegate>)delegate
                       keyHash:(EndpointKeyHash *)keyHash {
    if (delegate) {
        [[self.context getCallbackExecutor] addOperationWithBlock:^{
            [delegate onAttach:result resultContext:keyHash];
        }];
    }
}

- (void)notifyDetachedDelegate:(SyncResponseResultType)result delegate:(id<OnDetachEndpointOperationDelegate>)delegate {
    if (delegate) {
        [[self.context getCallbackExecutor] addOperationWithBlock:^{
            [delegate onDetach:result];
        }];
    }
}

- (void)onEndpointAccessTokenChanged {
    if (self.profileTransport) {
        [self.profileTransport sync];
    }
}

- (int)getRandomInteger {
    @synchronized (self) {
        return arc4random();
    }
}

@end
